### PR TITLE
Operator impl

### PR DESCRIPTION
**Pull request by ilackarms**
BOT NOTES: 
resolves https://github.com/solo-io/wasme/issues/42

### CODE REVIEW

**Comment by ilackarms** [comment]
/kick

added helm to ginkgo container

File: operator/happy_path_e2e_test.go [hunk 1]
```diff
@@ -0,0 +1,174 @@
...
+	b, err := ioutil.ReadFile(path)
+	if err != nil {
+		return err
+	}
+	extraArgs := []string{}
+	if ns != "" {
+		extraArgs = []string{"-n", ns}
+	}
+	return utils.KubectlDelete(b, extraArgs...)
+}
```
Comment (EItanya) [inline operator/happy_path_e2e_test.go]: These functions are essentially the same, can we make possibly combine them?

File: operator/install/kube/wasme/Chart.yaml [hunk 2]
```diff
@@ -0,0 +1,7 @@
+apiVersion: v1
+home: https://docs.solo.io/web-assembly-hub/latest
+icon: https://raw.githubusercontent.com/solo-io/wasme/master/docs/content/img/logo.png
+name: Wasme Operator
+sources:
+- https://github.com/solo-io/wasme
+version: v0.0.1
```
Comment (EItanya) [inline operator/install/kube/wasme/Chart.yaml]: In our other repos we have the directory structure `install/helm/<project>` Can we stick to the same here?

File: operator/install/kube/wasme/crds/wasme.io_v1_crds.yaml [hunk 3]
```diff
@@ -0,0 +1,23 @@
+apiVersion: apiextensions.k8s.io/v1beta1
+kind: CustomResourceDefinition
+metadata:
+  annotations:
+    helm.sh/hook: crd-install
```
Comment (EItanya) [inline operator/install/kube/wasme/crds/wasme.io_v1_crds.yaml]: This is a helm 2 concept, does not work with helm 3, is that intentional? I know we spoke about this briefly in the autopilot PR

File: pkg/operator/operator.go [hunk 4]
```diff
@@ -0,0 +1,205 @@
+package operator
```
Comment (EItanya) [inline pkg/operator/operator.go]: Can we add unit tests for this file?

File: Makefile [hunk 5]
```diff
@@ -117,8 +134,8 @@ endif
 .PHONY: publish-images
 publish-images:
 ifeq ($(RELEASE),"true")
-	docker push $(BUILDER_IMAGE):$(VERSION)
-	docker push $(CACHE_IMAGE):$(VERSION)
+	make wasme-image-push
+	make builder-image-push
```
[the remark this answers is not in the document]
Comment (ilackarms) [reply Makefile]: what do you mean do it with normal target?

File: operator/api/wasme/v1/filter_deployment.proto [hunk 6]
```diff
@@ -0,0 +1,113 @@
...
+message IstioDeploymentSpec {
+    // the kind of workload to deploy the filter to
+    // can either be Deployment or DaemonSet
+    string kind = 1;
+
+    // the name of the workload to deploy the filter to
+    // the workload must live in the same namespace as the FilterDeployment
+    // if empty, the filter will be deployed to all workloads in the namespace
+    string name = 2;
+}
```
[the remark this answers is not in the document]
Comment (ilackarms) [reply operator/api/wasme/v1/filter_deployment.proto]: yes i'd like to support labels. i can address in a follow-up. linking issue: 541b2c1155fffb15ccde92b8324f3e38f7339ba6

File: operator/generate.go [hunk 7]
```diff
@@ -0,0 +1,229 @@
...
+	v1 "k8s.io/api/core/v1"
+	rbacv1 "k8s.io/api/rbac/v1"
+	"k8s.io/apimachinery/pkg/api/resource"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"log"
+	"os"
+
+	"github.com/solo-io/autopilot/codegen"
+	"github.com/solo-io/autopilot/codegen/model"
+	"k8s.io/apimachinery/pkg/runtime/schema"
```
[the remark this answers is not in the document]
Comment (ilackarms) [reply operator/generate.go]: i'd like to hold off on adding this step in ci until a future PR

File: operator/install/kube/wasme/Chart.yaml (hunk 2 above)
Comment (ilackarms) [reply operator/install/kube/wasme/Chart.yaml]: the difference here is that wasme is not an operator-only project. `install` is associated only with the Operator, not the CLI version of wasme

File: operator/install/kube/wasme/crds/wasme.io_v1_crds.yaml (hunk 3 above)
Comment (ilackarms) [reply operator/install/kube/wasme/crds/wasme.io_v1_crds.yaml]: it is not required for helm 3, but i like leaving the helm 2 option available for now, as not everyone has made the switch.

for now, it isn't important as the helm chart is not currently released. it is only a way to generate the default installation manifest.

i would prefer to iterate on the helm chart when we are ready to begin publishing it

File: operator/install/kube/wasme/Chart.yaml (hunk 2 above)
Comment (EItanya) [reply operator/install/kube/wasme/Chart.yaml]: That is fine, I meant `operator/install/helm/<project>`

File: operator/generate.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (EItanya) [reply operator/generate.go]: Fine, can you just run formatting on the whole repo then

File: Makefile (hunk 5 above)
[the remark this answers is not in the document]
Comment (EItanya) [reply Makefile]: You are calling make from within the makefile, which feels like an anti pattern. I mean doing a target like this:
```
.PHONY: publish-images
publish-images: wasme-image-push builder-image-push
```

File: operator/install/kube/wasme/crds/wasme.io_v1_crds.yaml (hunk 3 above)
Comment (EItanya) [reply operator/install/kube/wasme/crds/wasme.io_v1_crds.yaml]: But we only have the crds in the `crds` folder, so it's not helm 2 compatible anyway. With new projects I am hesitant to support both. It is quite a bit of work, and not sure how much value it brings.
For now though I think it is fine until the feature is finished in autopilot

File: cloudbuild.yaml [hunk 8]
```diff
@@ -1,28 +1,35 @@
...
 
-  - name: 'gcr.io/$PROJECT_ID/go-mod-make:0.2.0'
+  - name: 'gcr.io/$PROJECT_ID/go-mod-make:0.2.1'
     args: ['build-images']
     env:
-      - 'GO111MODULE=on'
       - 'TAGGED_VERSION=$TAG_NAME'
-      - 'PROJECT_ROOT=github.com/solo-io/wasme'
-      - 'GOPATH=/workspace/gopath'
-      - 'GCLOUD_PROJECT_ID=$PROJECT_ID'
```
[the remark this answers is not in the document]
Comment (ilackarms) [reply cloudbuild.yaml]: fixed. `GCLOUD_PROJECT_ID=$PROJECT_ID` is now set inside the makefile

File: operator/install/kube/wasme/crds/wasme.io_v1_crds.yaml (hunk 3 above)
Comment (ilackarms) [reply operator/install/kube/wasme/crds/wasme.io_v1_crds.yaml]: if that's the case then i agree with removing it, but also agree that the change should happen in autopilot and we'll pick it up here eventually

File: Makefile (hunk 5 above)
[the remark this answers is not in the document]
Comment (ilackarms) [reply Makefile]: this is not considered an antipattern in makeake afaik. http://aggregate.org/rfisher/Tutorials/Make/make6.html and https://www.gnu.org/software/make/manual/html_node/Recursion.html are examples of using `make` recursively

we need the `make` here in order to support the conditional

**Pull request merged**